include/dynlib.h

```diff
@@ -0,0 +1,78 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2024-2024  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_DYNLIB_H
+#define DOSBOX_DYNLIB_H
+
+#include "std_filesystem.h"
+
+#ifdef WIN32
+
+#include <windows.h>
+#include <libloaderapi.h>
+
+using dynlib_handle = HINSTANCE;
+
+// Loads a dynamic-link library if it hasn't been opened yet, otherwise returns
+// a reference to it.
+inline dynlib_handle dynlib_open(const std_fs::path& path) noexcept
+{
+	return LoadLibraryA(path.string().c_str());
```

Comment:
Hmm, I'm not familiar with all this. I thought `std::string`s are just byte arrays, so they might contain UTF-8 or any random bytes. Then here we just take a pointer to the raw data with `c_str()` and pass it along. I don't see a problem?

My understanding is you only need to explicitly care about UTF-8 is you want to sort string, or calculate their lengths, or step through one glyph at a time, etc. But here we're just passing it along, no?

Reply:
`std::filesystem` is a bit special. Unless you tell it otherwise, it assumes the string you pass in to the constructor is the "native" type. 

On Unux like platform, passing in `char` is effectively `UTF-8`, or though it doesn't really matter.

On Windows, if you pass it a `char`, it treats it as a Windows "ANSI" string, which is basically whatever the current code page is. If you pass it a `wchar_t`, it treats it as a "wide" string (aka unicode in Windows terms).

However, AFAIK, if you explicitly tell `std::filesystem` that the path is UTF-8 using the `u8path()` static method in C++17 or the `char8_t` type in C++20, it will then convert the string to the native path type for the platform (eg: wchar_t) on Windows.

From the cppreference for `u8path()`:
> If path::value_type is char and native encoding is UTF-8, constructs a path directly as if by path(source) or path(first, last). Note: this is the typical situation of a POSIX system that uses Unicode, such as Linux.
> Otherwise, if path::value_type is wchar_t and native encoding is UTF-16 (this is the situation on Windows), or if path::value_type is char16_t (native encoding guaranteed UTF-16) or char32_t (native encoding guaranteed UTF-32), then first converts the UTF-8 character sequence to a temporary string tmp of type path::string_type and then the new path is constructed as if by path(tmp).
> Otherwise (for non-UTF-8 narrow character encodings and for non-UTF-16 wchar_t), first converts the UTF-8 character sequence to a temporary UTF-32-encoded string tmp of type [std::u32string](http://en.cppreference.com/w/cpp/string/basic_string), and then the new path is constructed as if by path(tmp) (this path is taken on a POSIX system with a non-Unicode multibyte or single-byte encoded filesystem).

Unfortunately, C++ then stuffed everything up by deprecating this function in favor of the new `char8_t` type...

Reply:
So, TLDR, on Windows, we should be creating `wchar_t` paths (either directly or through UTF-8 conversion), and then use the `W` version of the Win32 API functions.

Reply:
Thanks for the explanation @shermp, now it makes sense. This is a rather nasty Windows bug. I expect people on non-English Windows to have UTF-8 characters in path names.

Possibly related to https://github.com/dosbox-staging/dosbox-staging/issues/2681 ?

Would you be interested in fixing this?

Reply:
Yeah, definitely related to that issue.

I've thought about it quite a bit, but I can have a go at some point when I have time. There's just sooo many places in dosbox where we are touching the host filesystem though!

It's not just the `*A` vs `*W` functions either. `fopen` is also affected

Reply:
@shermp Looks like you know what to do, so any initial implementation would be appreciated even if it's not complete.

I think the best cure for the "OMG this task is too big, I don't even where to start... so I won't" 😆  type of mental block is to pick a small and achievable target. Maybe just pick one module or even just a single file thst interests you, and only swap out the calls there. If that means duplicating some of the "wrong" path handling functions so the "correct" and "wrong" variants can coexis t, so be it!

A stepwise refactoring approach like this is much safer and way better for big refactoring jobs if it's feasible. Changes are safer, testing gets simpler, reviewing becomes less of a burden, so overall stress levels go down! So yeah, if you wanna give it a go, just fix _one_ user-observable bug with the minimum code change first, and we'll see where that takes us as a team.